src/main/java/cubicchunks/lighting/FirstLightProcessor.java

```diff
@@ -60,155 +56,198 @@
+		public int hashCode(int e) {
 			return e;
 		}
 
-		@Override public boolean equals(int a, int b) {
+		@Override
+		public boolean equals(int a, int b) {
 			return a == b;
 		}
 	};
-	//mutableBlockPos variable to avoid creating thousands of instances of BlockPos
-	private BlockPos.MutableBlockPos mutablePos = new BlockPos.MutableBlockPos();
-	private ICubeCache cache;
 
+
+	private final MutableBlockPos mutablePos = new MutableBlockPos();
+
+	private final ICubeCache cache;
+
+
+	/**
+	 * Creates a new FirstLightProcessor for the given world.
+	 *
+	 * @param world the world for which the FirstLightProcessor will be used
+	 */
 	public FirstLightProcessor(ICubicWorld world) {
 		this.cache = world.getCubeCache();
 	}
 
+
 	/**
-	 * Sets blocks exposed to sunlight to full brightness.
+	 * Initializes skylight in the given cube. The skylight will be consistent with respect to the world configuration
+	 * and already existing cubes. It is however possible for cubes being considered lit at this stage to be occluded
+	 * by cubes being generated further up.
 	 *
-	 * This method should NOT set blocks to dark.
-	 *
-	 * This method should be called as soon as a cube is filled with blocks. It updates only that one cube.
+	 * @param cube the cube whose skylight is to be initialized
 	 */
-	public void earlySkylightMap(Cube cube) {
-		Column column = cube.getColumn();
+	public void initializeSkylight(Cube cube) {
+		IOpacityIndex opacityIndex = cube.getColumn().getOpacityIndex();
 
-		CubeCoords cubePos = cube.getCoords();
-		cubePos.forEachBlockPosMutableTopDown((pos)->{
-			int topY = getHeightmapValue(column, blockToLocal(pos.getX()), blockToLocal(pos.getZ()));
-			if(pos.getY() <= topY) {
-				return false;
+		int cubeMinY = Coords.cubeToMinBlock(cube.getY());
+
+		for (int localX = 0; localX < Cube.SIZE - 1; ++localX) {
+			for (int localZ = 0; localZ < Cube.SIZE - 1; ++localZ) {
+				for (int localY = Cube.SIZE - 1; localY >= 0; --localY) {
+
+					if (opacityIndex.isOccluded(localX, cubeMinY + localY, localZ)) {
+						break;
+					}
+
+					cube.setSkylight(localX, localY, localZ, 15);
+				}
 			}
-			cube.setLightFor(EnumSkyBlock.SKY, pos, 15);
-			return true;
-		});
+		}
 	}
 
-	public boolean diffuseSkylight(Cube generatedCube) {
-		ICubicWorld world = generatedCube.getWorld();
+	/**
+	 * Diffuses skylight in the given cube and all cubes affected by this update.
+	 *
+	 * @param cube the cube whose skylight is to be initialized
+	 */
+	public void diffuseSkylight(Cube cube) {
+		ICubicWorld world = cube.getWorld();
 
-		//cache min/max Y, generating them may be expensive
+		// Cache min/max Y, generating them may be expensive
 		int[][] minBlockYArr = new int[16][16];
 		int[][] maxBlockYArr = new int[16][16];
 
-		int minBlockX = cubeToMinBlock(generatedCube.getX());
-		int maxBlockX = cubeToMaxBlock(generatedCube.getX());
+		int minBlockX = Coords.cubeToMinBlock(cube.getX());
+		int maxBlockX = Coords.cubeToMaxBlock(cube.getX());
```

Comment:
Renaming generatedCube to cube makes it conform to the overall standard in this class. This class itself does not impose a restriction on diffuseSkylight being called for generatedCubes only.
I'll put the static import back in, personal preference of mine not to use them, as they blur the line of which methods are implemented in which class.